src/hoisted.jl

```diff
@@ -0,0 +1,84 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+addbyte!(x::Int, c) = x + 1
+addbyte!(x::Vector{UInt8}, c) = push!(x, c)
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start.
+    s = ps.s
+    len = parse_string(ps, 0)
+
+    # Now read the string itself
+    ps.s = s
+    ret = Compat.UTF8String(UInt8[])
+    b = sizehint!(ret.data, len)
+    parse_string(ps, b)
+    ret
```

Comment:
Oh, right.   You would have to:
```jl
ccall(:jl_array_to_string, Ref{String}, (Any,), b)
```
and calling an undocumented C API function doesn't seem any better than mutating the `String` internals.

(There should probably be an `unsafe_array_to_string` function for this.)